worker/lease/manager.go

```diff
@@ -378,55 +425,98 @@ func (manager *Manager) nextTick(lastTick time.Time) <-chan time.Time {
 		}
 		nextTick = info.Expiry
 	}
-	manager.config.Logger.Tracef("[%s] next tick decided on %v %v", manager.logContext, nextTick.Sub(now), nextTick)
-	return clock.Alarm(manager.config.Clock, nextTick)
+	manager.config.Logger.Tracef("[%s] next expire decided on %v %v",
+		manager.logContext, nextTick.Sub(now).Round(time.Millisecond), nextTick)
+	manager.setNextTimeout(nextTick)
+}
+
+func (manager *Manager) setupInitialTimer() {
+	// Create a timer at max timeout, we'll update after refreshing leases
+	manager.muNextTimeout.Lock()
+	manager.timer = manager.config.Clock.NewTimer(manager.config.MaxSleep)
+	manager.muNextTimeout.Unlock()
+	// lastTick has never happened, so pass in the epoch time
+	manager.computeNextTimeout(time.Time{}, manager.config.Store.Leases())
+}
+
+func (manager *Manager) setNextTimeout(t time.Time) {
+	manager.muNextTimeout.Lock()
+	manager.nextTimeout = t
+	d := t.Sub(manager.config.Clock.Now())
+	manager.timer.Reset(d)
+	manager.muNextTimeout.Unlock()
 }
 
-// retryingTick runs tick and retries any timeouts.
-func (manager *Manager) retryingTick(now time.Time) {
-	manager.config.Logger.Tracef("[%s] tick looking for leases to expire\n", manager.logContext)
+// ensureNextTimeout makes sure that the next timeout happens no-later than
+// duration from now.
+func (manager *Manager) ensureNextTimeout(d time.Duration) {
+	manager.muNextTimeout.Lock()
+	next := manager.nextTimeout
+	proposed := manager.config.Clock.Now().Add(d)
+	if next.After(proposed) {
+		manager.config.Logger.Tracef("[%s] ensuring we wake up before %v at %v\n",
+			manager.logContext, d, next)
+		manager.nextTimeout = next
```

Comment:
Should `proposed` be used here and in the log line above?

Reply:
wow, you're absolutely right, and clearly we have a missing test as it didn't catch this.
I'll work on a test for this.